test/boot_raw_fallback.c

```diff
@@ -0,0 +1,414 @@
+	gint err_code;
+
+	gsize image_size;
+	guint64 region_start;
+	guint64 region_size;
+} BootRawFallbackData;
+
+static gboolean process_file(const gchar *filepath,
+		gsize offset,
+		gsize size,
+		gchar fill,
+		gboolean random,
+		gboolean compare)
+{
+	GError *error = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(GFileIOStream) stream = NULL;
+	g_autofree gchar *buf = NULL;
+	g_autofree gchar *cmp_buf = NULL;
+	const gsize buf_size = 4096;
+	gboolean res;
+
+	file = g_file_new_for_path(filepath);
+	stream = g_file_open_readwrite(file, NULL, &error);
+	g_assert_no_error(error);
+	g_assert_nonnull(stream);
+
+	res = g_seekable_seek(G_SEEKABLE(stream), offset, G_SEEK_SET, NULL, &error);
+	g_assert_no_error(error);
+	g_assert_true(res);
+
+	if (!random) {
+		buf = g_malloc(buf_size);
+		memset(buf, fill, buf_size);
+	}
+
+	if (compare)
+		cmp_buf = g_malloc(buf_size);
+
+	while (size) {
+		gsize bytes_to_process = size < buf_size ? size : buf_size;
+		gsize bytes_processed;
+
+		if (random)
+			buf = (gchar*) random_bytes(bytes_to_process, 31415);
+
+		if (compare) {
+			res = g_input_stream_read_all(g_io_stream_get_input_stream(G_IO_STREAM(stream)),
+					cmp_buf,
+					bytes_to_process,
+					&bytes_processed,
+					NULL,
+					&error);
+
+			if (memcmp(buf, cmp_buf, bytes_processed)) {
+				return FALSE;
+			}
+		} else {
+			res = g_output_stream_write_all(g_io_stream_get_output_stream(G_IO_STREAM(stream)),
+					buf,
+					bytes_to_process,
+					&bytes_processed,
+					NULL,
+					&error);
+		}
+		g_assert_no_error(error);
+		g_assert_true(res);
+
+		size -= bytes_processed;
+	}
+	return TRUE;
+}
+
+static gboolean fill_file(const gchar *filepath,
+		gsize offset,
+		gsize size,
+		gchar fill,
+		gboolean random)
+{
+	return process_file(filepath, offset, size, fill, random, FALSE);
+}
+
+static gboolean compare_file(const gchar *filepath,
+		gsize offset,
+		gsize size,
+		gchar fill,
+		gboolean random)
+{
+	return process_file(filepath, offset, size, fill, random, TRUE);
+}
+
+static void boot_raw_fallback_fixture_set_up(BootRawFallbackFixture *fixture,
+		gconstpointer user_data)
+{
+	fixture->tmpdir = g_dir_make_tmp("rauc-XXXXXX", NULL);
+	g_assert_nonnull(fixture->tmpdir);
+}
+
+static void boot_raw_fallback_fixture_tear_down(BootRawFallbackFixture *fixture,
+		gconstpointer user_data)
+{
+	if (!fixture->tmpdir)
+		return;
+
+	g_assert(test_rmdir(fixture->tmpdir, "") == 0);
+}
+
+static RaucImage* create_source_image(const gchar *dirname,
+		const gchar *imagename,
+		gsize size)
+{
+	RaucImage *image;
+	g_autofree gchar *imagepath;
+
+	imagepath = g_build_filename(dirname, imagename, NULL);
+
+	image = g_new0(RaucImage, 1);
+	image->slotclass = g_strdup("rootfs");
+	image->filename = g_strdup(imagepath);
+	image->checksum.size = size;
+	image->checksum.digest = g_strdup("0xdeadbeef");
+
+	g_assert(test_prepare_dummy_file(dirname, imagename, size, "/dev/zero") == 0);
+
+	fill_file(imagepath, 0, size, 0x00, TRUE);
```

Comment:
This could be replaced with `write_random_file()` from `test/common.c`.

Reply:
True. But as it is important that the content meets the expectations from compare_file(), I would like to stick with fill_file()-compare_file() pairs.